packages/shared/src/hooks/useCompanionTrigger.tsx

```diff
@@ -0,0 +1,160 @@
+
+import { useExtensionPermission } from './useExtensionPermission';
+import { useLazyModal } from './useLazyModal';
+import { LazyModal } from '../components/modals/common/types';
+import { useActions } from './useActions';
+import { ActionType } from '../graphql/actions';
+
+type CompanionTriggerProps = {
+  post: Post;
+  index?: number;
+  row?: number;
+  column?: number;
+};
+
+type OpenArticle = (data?: CompanionTriggerProps) => () => Promise<void>;
+
+export type CompanionTrigger = {
+  onFeedArticleClick: FeedPostClick;
+  onPostArticleClick: PostClick;
+};
+
+type PostClick = (e: React.MouseEvent, data: { post: Post }) => Promise<void>;
+
+export default function useCompanionTrigger(
+  customPostClickHandler: (
+    post: Post,
+    index?: number,
+    row?: number,
+    column?: number,
+  ) => Promise<void>,
+): CompanionTrigger {
+  const isExtension = process.env.TARGET_BROWSER;
+  const { trackEvent } = useContext(AnalyticsContext);
+  const { closeModal: closeLazyModal, openModal: openLazyModal } =
+    useLazyModal();
+  const { isActionsFetched, checkHasCompleted, completeAction } = useActions();
+  const gb = useGrowthBook();
+
+  const { requestContentScripts, useContentScriptStatus } =
+    useExtensionPermission({
+      origin: 'companion modal permission button',
+    });
+
+  const { contentScriptGranted } = useContentScriptStatus();
+  const alreadyCompleted = checkHasCompleted(
+    ActionType.EngagementLoopJuly2023CompanionModal,
+  );
+
+  const handleCompleteAction = useCallback(() => {
+    if (isActionsFetched)
+      completeAction(ActionType.EngagementLoopJuly2023CompanionModal);
+  }, [isActionsFetched, completeAction]);
+
+  const closeModal = useCallback(() => {
+    closeLazyModal();
+    handleCompleteAction();
+
+    trackEvent({
+      event_name: `close companion permission popup`,
+      extra: JSON.stringify({ origin: 'automatic' }),
+    });
+  }, [closeLazyModal, handleCompleteAction, trackEvent]);
+
+  const activateCompanion = useCallback(async () => {
+    await requestContentScripts();
+    closeModal();
+    handleCompleteAction();
+  }, [requestContentScripts, closeModal, handleCompleteAction]);
+
+  const openArticle: OpenArticle = useCallback(
+    (data) => async () => {
+      await customPostClickHandler(
+        data.post,
+        data.index,
+        data.row,
+        data.column,
+      );
+
+      closeModal();
+      handleCompleteAction();
+    },
+    [customPostClickHandler, closeModal, handleCompleteAction],
+  );
+
+  const openModal = useCallback(
+    (data?: CompanionTriggerProps) => {
+      openLazyModal({
+        type: LazyModal.CompanionModal,
+        props: {
+          url: data?.post?.permalink,
+          onReadArticleClick: openArticle(data),
+          onActivateCompanion: activateCompanion,
+        },
+      });
+
+      trackEvent({
+        event_name: `open companion permission popup`,
+        extra: JSON.stringify({ origin: 'automatic' }),
```

Comment:
Same comment as above on the type, please take careful not of the taxonomy sheet